homeassistant/components/conversation/default_agent.py

```diff
@@ -683,45 +703,60 @@ def _recognize(
 
         return maybe_result
 
-    def _get_all_entity_names(self) -> TextSlotList:
-        """Get slot list with all entity names in Home Assistant."""
-        if self._all_entity_names is not None:
-            return self._all_entity_names
+    def _get_unexposed_entity_names(self, text: str) -> TextSlotList:
+        """Get filtered slot list with unexposed entity names in Home Assistant."""
+        if self._unexposed_names_trie is None:
+            # Build trie
+            self._unexposed_names_trie = Trie()
+            entity_registry = er.async_get(self.hass)
 
-        entity_registry = er.async_get(self.hass)
-        all_entity_names: list[tuple[str, str, dict[str, Any]]] = []
+            for state in self.hass.states.async_all():
+                if async_should_expose(self.hass, DOMAIN, state.entity_id):
+                    continue
 
-        for state in self.hass.states.async_all():
-            context = {"domain": state.domain}
-            if state.attributes:
-                # Include some attributes
-                for attr in DEFAULT_EXPOSED_ATTRIBUTES:
-                    if attr not in state.attributes:
-                        continue
-                    context[attr] = state.attributes[attr]
+                context = {"domain": state.domain}
+                if state.attributes:
+                    # Include some attributes
+                    for attr in DEFAULT_EXPOSED_ATTRIBUTES:
+                        if attr not in state.attributes:
+                            continue
+                        context[attr] = state.attributes[attr]
```

Comment:
Can't we share this code with `_make_slot_lists` ?

Reply:
Refactored to share code